src/System.Private.CoreLib/shared/System/Text/ASCIIEncoding.cs

```diff
@@ -248,673 +422,341 @@ public override unsafe int GetCharCount(byte[] bytes, int index, int count)
-
-            // Assert because we shouldn't be able to have a null encoder.
-            Debug.Assert(encoderFallback != null, "[ASCIIEncoding.GetByteCount]Attempting to use null fallback encoder");
-
-            char charLeftOver = (char)0;
-            EncoderReplacementFallback fallback = null;
-
-            // Start by assuming default count, then +/- for fallback characters
-            char* charEnd = chars + charCount;
-
-            // For fallback we may need a fallback buffer, we know we aren't default fallback.
-            EncoderFallbackBuffer fallbackBuffer = null;
-            char* charsForFallback;
-
-            if (encoder != null)
             {
-                charLeftOver = encoder._charLeftOver;
-                Debug.Assert(charLeftOver == 0 || char.IsHighSurrogate(charLeftOver),
-                    "[ASCIIEncoding.GetByteCount]leftover character should be high surrogate");
-
-                fallback = encoder.Fallback as EncoderReplacementFallback;
-
-                // We mustn't have left over fallback data when counting
-                if (encoder.InternalHasFallbackBuffer)
-                {
-                    // We always need the fallback buffer in get bytes so we can flush any remaining ones if necessary
-                    fallbackBuffer = encoder.FallbackBuffer;
-                    if (fallbackBuffer.Remaining > 0 && encoder._throwOnOverflow)
-                        throw new ArgumentException(SR.Format(SR.Argument_EncoderFallbackNotEmpty, this.EncodingName, encoder.Fallback.GetType()));
-
-                    // Set our internal fallback interesting things.
-                    fallbackBuffer.InternalInitialize(chars, charEnd, encoder, false);
-                }
-
-                // Verify that we have no fallbackbuffer, for ASCII its always empty, so just assert
-                Debug.Assert(!encoder._throwOnOverflow || !encoder.InternalHasFallbackBuffer ||
-                    encoder.FallbackBuffer.Remaining == 0,
-                    "[ASCIICodePageEncoding.GetByteCount]Expected empty fallback buffer");
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.bytes, ExceptionResource.ArgumentOutOfRange_IndexCountBuffer);
             }
-            else
+
+            if ((uint)charIndex > (uint)chars.Length)
             {
-                fallback = this.EncoderFallback as EncoderReplacementFallback;
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.charIndex, ExceptionResource.ArgumentOutOfRange_Index);
             }
 
-            // If we have an encoder AND we aren't using default fallback,
-            // then we may have a complicated count.
-            if (fallback != null && fallback.MaxCharCount == 1)
+            // Using GetRawSzArrayData allows us to get a usable reference for empty
+            // arrays so we don't need to special-case the null value.
+
+            fixed (byte* pBytes = &JitHelpers.GetRawSzArrayData(bytes))
```

Comment:
These JitHelpers do not exist in CoreRT (not sure about Mono). We should leave introducing them into a separate change, I think.

Reply:
Looks like I can rewrite this as `bytes.GetRawSzArrayData()`, which should pick up [the instance method](https://github.com/dotnet/corert/blob/50d9f23cb9fac7dfb7620f75debf5654a6751de4/src/System.Private.CoreLib/src/System/Array.CoreRT.cs#L230) on CoreRT and [the `JitHelpers` extension method](https://github.com/dotnet/coreclr/blob/8b3c55c36531689c337df37f15eb5a40530fa42f/src/System.Private.CoreLib/src/System/Runtime/CompilerServices/jithelpers.cs#L105) on coreclr.

Reply:
Yes, it may be possible; but I think it would be better to look at this optimization separately. It can be applied wholesale to pretty much all `fixed` statements across CoreCLR/CoreFX.

Also, when we have discussed pinning behavior for Span, the security folks insisted on that it is important to normalize empty buffer to null pointer for defense in depth. I think that it is unfortunate and I do not think that it provides much defense in depth; but it would be useful to get everybody on the same page. If we are ok with using this in "regular" framework code like encodings ourselves, we should have the API to allow us customers to do it too.

Reply:
> Also, when we have discussed pinning behavior for Span, the security folks insisted on that it is important to normalize empty buffer to null pointer for defense in depth.

IIRC the problem was that devs can write the entirely safe code `var value = mySpan.GetPinnableReference();`, and now it's dereferencing arbitrary data that might cause an AV or which you should otherwise not have access to. By forcing such "unsafe" accesses to be bounced through `MemoryMarshal`, we're essentially forcing the developer to specify their intent to retrieve a reference which might not be safely derefable.

I would have no problems at all with the syntax `fixed (T* ptr = &theSpan)` putting a non-null pointer into the _ptr_ local, even for empty spans. (I'd also have no problems at all with empty T[] doing the same thing, but this would be a breaking change.) Once pointers are in play, it's up to the dev to be responsible with them. But I don't think it's possible to make this scenario work without also making `GetPinnableReference()` above return a dangerous ref, so that's the world we're in.

See also what a wise man once said over at https://github.com/dotnet/corefx/issues/23881#issuecomment-352165338. 😉

Reply:
Anyway, `JitHelpers` aside, I can easily revert those lines. As you suggested, we can address it as part of a larger sweep if we're so inclined.

Reply:
> `fixed (T* ptr = &theSpan)` putting a non-null pointer into the `ptr` local, even for empty spans.

I agree with that, but that is not the design that was choosen because of security concerns and array prior art. And this design in combination with `readonly` made `GetPinnableReference` perfectly safe function.

> I can easily revert those lines.

I would prefer that at least until we have a helper that takes `T[]` and returns `ref T`.

Reply:
> I would prefer that at least until we have a helper that takes T[] and returns ref T.

Sure, no problem! I'd recommend continuing this discussion at https://github.com/dotnet/coreclr/pull/22917 or https://github.com/dotnet/corefx/issues/33706 if there's interest in having such an API exposed publicly.